ADR/0018-apps-continuous-perf-testing.md

```diff
@@ -0,0 +1,72 @@
+>
+> If any of these metrics does not align to historical results, we mark this test result as a failure.
+
+And even if some test fails, the application team should be able to review the data and change the test definition, so next time a new test result is being evaluated based on the historical results, this new result will be one of them.
+
+The algorithm that says if the current test passes or fails when compared to historical results can vary. It can be:
+
+- Checking minimum and maximum of the metric in historical data and making sure the new result is in this range.
+- Using a floating average of historical data and setting a safe range around it and making sure the new result is in that range.
+- Some machine learning thingy.
+
+Goal of this ADR is to propose a way for this kind of testing to be implemented in Stonesoup (feature [STONE-679](https://issues.redhat.com/browse/STONE-679)). Even if it would not deprecate full-blown performance and scale testing, having a release gate with some short and small scale performance test is desirable for many application teams.
+
+We are considering two different options here. They differ in what portion of logic mentioned above is executed where.
+
+### Glossary
+
+- SUT: Subject Under Test
+- Horreum: a results storage service: <https://github.com/Hyperfoil/Horreum>
+- CPT: Continuous Performance Testing
+
+## Decision
+
+Let's use this architecture with a single Horreum instance that would be managed by Red Hat and used by all customers.
+
+![Architecture diagram with Horreum](assets/0018-apps-continuous-perf-testing.svg "Architecture diagram with Horreum")
+
+1. Performance test runs in the Tekton pipeline and generates JSON with test parameters and results.
+2. We gather configured monitoring metrics and add them to the JSON.
+3. We upload the JSON with all the data to Horreum.
+4. Horreum performs result analysis, looking for changes in configured metrics.
+4. We get PASS/FAIL decision from Horreum back to the pipeline, so pipeline can rerurn proper result.
+
+## Consequences
+
+Pros:
+
+- We can provide some best practices to customers wanting to do some basic performance regression testing.
+- Horreum is end-to-end solution that already exists, is already used by multiple teamsn and has documentation for the whole process.
+- StoneSoup UI can display all the trends because historical data will be available in Horreum.
+- Red Hat team that develops Horreum is willing to help with this effort.
+- Integration scripts that would be running in the pipeline (to gather monitoring data, to upload results to Horreum, to get PASS/FAIL decision from Horreum...) mostly exists.
+
+Cons:
+
+- Somebody has to manage the Horreum instance. Required development time on it.
+- Although Horreum uses Keycloak as well, some changes in the Horreum authentication mechanism might be required.
+- Maybe also some scaling concerns of Horreum.
```

Comment:
What does this mean? How concerned should I be with this point? Is there a threshold where this starts to become _more_ of a concern?

Reply:
As of now, I know multiple teams are using Horreum and it is far from it's full capacity. But to be fair, there was no propper perfomance&scale testing of Horreum itself as far as I know, so we really do not know what is the actual scale Horreum can operate at. It is same for StoneSoup itself (I'm part of the team that tries to perf&scale test StoneSoup itself - see STONE-224). So the answer is "I do not know" as of now.

Reply:
We can be more descriptive here then -- mentioning the specific scale concerns (i.e. untested). This might give credence to having multiple Horreum instances, one per tenant.

Reply:
I see! Great. Changed the wording to mention that.

Reply:
Are there pushed changes or am I looking in the wrong place? I do not see any change in the wording.

Reply:
Nevermind, I was looking at an old commit. THanks.